pkg/processor/eventprocessor/pool.go

```diff
@@ -0,0 +1,193 @@
+	for _, object := range objects {
+		newFixedPool.objectsChan <- object
+	}
+
+	return newFixedPool
+}
+
+func (sa *syncPoolAllocator) Allocate(timeout time.Duration) (EventProcessor, error) {
+	// TODO: think about reworking of atomic operations logic as it might affect performance
+	// we don't want to completely lock here, but we'll use atomic to inc counters where possible
+	atomic.AddUint64(&sa.statistics.AllocationCount, 1)
+
+	// get total number of objects
+	totalNumberObjects := len(sa.objects)
+	currentNumberOfAvailableObjects := len(sa.objectsChan)
+	percentageOfAvailableObjects := float64(currentNumberOfAvailableObjects*100.0) / float64(totalNumberObjects)
+
+	// measure how many objects are available in the queue while we're allocating
+	atomic.AddUint64(&sa.statistics.AllocationObjectsAvailablePercentage, uint64(percentageOfAvailableObjects))
+
+	// try to allocate a worker and fall back to default immediately if there's none available
+	select {
+	case workerInstance := <-sa.objectsChan:
+		atomic.AddUint64(&sa.statistics.AllocationSuccessImmediateTotal, 1)
+
+		return workerInstance, nil
+	default:
+
+		// if there's no timeout, return now
+		if timeout == 0 {
+			atomic.AddUint64(&sa.statistics.AllocationTimeoutTotal, 1)
+			return nil, ErrNoAvailableObjects
+		}
+
+		waitStartAt := time.Now()
+
+		// if there is a timeout, try to allocate while waiting for the time
+		// to pass
+		select {
+		case workerInstance := <-sa.objectsChan:
+			atomic.AddUint64(&sa.statistics.AllocationSuccessAfterWaitTotal, 1)
+			atomic.AddUint64(&sa.statistics.AllocationWaitDurationMilliSecondsSum,
+				uint64(time.Since(waitStartAt).Nanoseconds()/1e6))
+			return workerInstance, nil
+		case <-time.After(timeout):
+			atomic.AddUint64(&sa.statistics.AllocationTimeoutTotal, 1)
+			return nil, ErrNoAvailableObjects
+		}
+	}
+}
+
+func (sa *syncPoolAllocator) Release(object EventProcessor) {
+	sa.objectsChan <- object
+}
+
+func (sa *syncPoolAllocator) Shareable() bool {
+	return true
+}
+
+func (sa *syncPoolAllocator) GetObjects() []EventProcessor {
+	return sa.objects
+}
+
+func (sa *syncPoolAllocator) GetNumObjectsAvailable() int {
+	return len(sa.objectsChan)
+}
+
+func (sa *syncPoolAllocator) SetObjects(objects []EventProcessor) error {
+	sa.objects = objects
+	return nil
+}
+
+// GetStatistics returns object allocator statistics
+func (sa *syncPoolAllocator) GetStatistics() *AllocatorStatistics {
+	return &sa.statistics
+}
+
+func (sa *syncPoolAllocator) SignalDraining() error {
+	errGroup, _ := errgroup.WithContext(context.Background(), sa.logger)
+
+	for _, objectInstance := range sa.GetObjects() {
+		objectInstance := objectInstance
+
+		errGroup.Go(fmt.Sprintf("Drain object %d", objectInstance.GetIndex()), func() error {
+			// if object is not already drained, signal it to drain events
+			if err := objectInstance.Drain(); err != nil {
+				return errors.Wrapf(err, "Failed to signal object %d to drain events", objectInstance.GetIndex())
```

Comment:
the errors look weird now with `"object"`, but not sure we want to keep `worker` here..

Reply:
@TomerShor I think it's fine, worker is something internal as well as object, so it's more for us than for users